pkg/dockerregistry/client.go

```diff
@@ -43,18 +47,57 @@ type client struct {
 	connections map[string]connection
 }
 
+// Connect accepts the name of a registry in the common form Docker provides and will
+// create a connection to the registry. Callers may provide a host, a host:port, or
+// a fully qualified URL. When not providing a URL, the default scheme will be "https"
 func (c *client) Connect(name string) (Connection, error) {
-	if len(name) == 0 {
-		name = "index.docker.io"
+	target, err := normalizeRegistryName(name)
+	if err != nil {
+		return nil, err
 	}
-	if conn, ok := c.connections[name]; ok {
+	prefix := target.String()
+	if conn, ok := c.connections[prefix]; ok {
 		return conn, nil
 	}
-	conn := newConnection(name)
-	c.connections[name] = conn
+	conn := newConnection(*target)
+	c.connections[prefix] = conn
 	return conn, nil
 }
 
+func normalizeRegistryName(name string) (*url.URL, error) {
+	prefix := name
+	if len(prefix) == 0 {
+		prefix = "index.docker.io"
+	}
+	switch {
+	case strings.HasPrefix(prefix, "http://"), strings.HasPrefix(prefix, "https://"):
```

Comment:
Regex are bad because no one understands regex, avoid using them in general for anything that can be more clearly defined otherwise.  Performance is important but clarity is more so.  

Go generally prefers switch over multiple conditionals, but a simple Boolean as you mentioned is simpler - I'll make that change.

The coding style I think we should be trying for is when someone familiar with go looks at this code 2 years hence, would they mistake the intent or the correctness of the code.  Verbosity can be bad, succinctness can be bad.

> On Apr 19, 2015, at 12:42 PM, Ben Parees <notifications@github.com> wrote:
> 
> In pkg/dockerregistry/client.go:
> 
> >  	return conn, nil
> >  }
> >  
> > +func normalizeRegistryName(name string) (*url.URL, error) {
> > +	prefix := name
> > +	if len(prefix) == 0 {
> > +		prefix = "index.docker.io"
> > +	}
> > +	switch {
> > +	case strings.HasPrefix(prefix, "http://"), strings.HasPrefix(prefix, "https://"):
> i don't know that i agree with the definition of "better" here, but even if you don't a regex (for performance or readability concerns? i comment would handle the latter), I would also say that:
> if !a && !b {
> prepend
> }
> 
> would be better than this switch as it's currently written, since what you've currently written is semantically:
> 
> if a || b {
> // do nothing
> } else {
> prepend()
> }
> which seems like an ineloquent (and needlessy verbose) expression.
> 
> —
> Reply to this email directly or view it on GitHub.
>